### ISSUE TITLE
TSAN/MSAN fails with vm.mmap_rnd_bits=32 even with 18.1.3

### ISSUE THREAD

**Issue by Sjors** [bitcoin/bitcoin#30674]
The Cirrus CI on my fork of the repo runs on Ubuntu 24.04 with kernel version 6.8.0-38. This has `vm.mmap_rnd_bits=32` set, which causes the TSAN and MSAN jobs to fail.

See:

TSAN: https://cirrus-ci.com/task/6619444124844032

```
FAIL: minisketch/test
=====================
ThreadSanitizer: CHECK failed: tsan_platform_linux.cpp:282 "((personality(old_personality | ADDR_NO_RANDOMIZE))) != ((-1))" (0xffffffffffffffff, 0xffffffffffffffff) (tid=42931)
FAIL minisketch/test (exit status: 139)
FAIL: univalue/test/object
==========================
ThreadSanitizer: CHECK failed: tsan_platform_linux.cpp:282 "((personality(old_personality | ADDR_NO_RANDOMIZE))) != ((-1))" (0xffffffffffffffff, 0xffffffffffffffff) (tid=42964)
FAIL univalue/test/object (exit status: 139)
FAIL: qt/test/test_bitcoin-qt
=============================
ThreadSanitizer: CHECK failed: tsan_platform_linux.cpp:282 "((personality(old_personality | ADDR_NO_RANDOMIZE))) != ((-1))" (0xffffffffffffffff, 0xffffffffffffffff) (tid=42994)
FAIL qt/test/test_bitcoin-qt (exit status: 139)
```

MSAN: https://cirrus-ci.com/task/4578750543691776

```
unning tests: base58_tests from test/base58_tests.cpp
Running tests: base64_tests from test/base64_tests.cpp
MemorySanitizer: CHECK failed: msan_linux.cpp:192 "((personality(old_personality | ADDR_NO_RANDOMIZE))) != ((-1))" (0xffffffffffffffff, 0xffffffffffffffff) (tid=22112)
    <empty stack>
make[3]: *** [Makefile:22563: test/base32_tests.cpp.test] Error 1
make[3]: *** Waiting for unfinished jobs....
MemorySanitizer: CHECK failed: msan_linux.cpp:192 "((personality(old_personality | ADDR_NO_RANDOMIZE))) != ((-1))" (0xffffffffffffffff, 0xffffffffffffffff) (tid=22137)
    <empty stack>
```

This job was from mid July. Just in case I reproduced it against todays master: https://github.com/Sjors/bitcoin/pull/57

My (limited) understanding is that the underlying issue should have been fixed and the fix has been backported to llvm 18.1.3: https://github.com/google/sanitizers/issues/1614#issuecomment-2010316781 

Ubuntu 24.04 has shipped that version since early July:https://launchpad.net/ubuntu/noble/amd64/clang-18

I can see in the CI log this this version was indeed used:

```
Get:123 http://archive.ubuntu.com/ubuntu noble-updates/main amd64 libllvm18 amd64 1:18.1.3-1ubuntu1 [27.5 MB]
```

Although I can trivially work around the issue by setting `vm.mmap_rnd_bits=28`, perhaps there is a deeper issue worth investigating.

Possibly related: https://github.com/ClickHouse/ClickHouse/issues/64086 (they also tried 18.1.3 and 18.1.6).

**Comment by Sjors** [comment]
@maflcko yes, I first reproduced the issue and then tested the workaround `vm.mmap_rnd_bits=28`.

I'll try those clang-19 PRs.

**Comment by fanquake** [comment]
Has started happening again in the TSAN CI (Clang 19.1.4)? https://cirrus-ci.com/task/5338932714405888?logs=ci#L2213

**Comment by Sjors** [comment]
When I install the ARM version of Ubuntu in VM on an M4 macOS machine using UTM, the default of `vm.mmap_rnd_bits` is 33. Though I haven't checked if the native ARM job cares.

**Comment by willcl-ark** [comment]
Please update the CI readme.md with the needed info, thanks.

**Comment by maflcko** [comment]
for reference, the workaround isn't needed for the GHA runners, but it is needed locally and for the Cirrus runners, see also the workflow yaml: https://github.com/bitcoin/bitcoin/blob/c862936d16a640690ef4c89456738fe4bb99be54/.github/workflows/ci.yml#L554-L557

**Comment by Sjors** [comment]
FWIW it's not really fixed, #33674 just documents the workaround instruction. But with the move to Cirrus runners (and its Github CI fallback) in #32989 I'm no longer running self-hosted CI on my fork(s), so I no longer care about the issue.

**Comment by maflcko** [comment]
> FWIW it's not really fixed, [#33674](https://github.com/bitcoin/bitcoin/pull/33674) just documents the workaround instruction.

I don't think another fix is possible, as explained in the linked llvm commit above:

> It is difficult to make the
mappings any larger, given the 44-bit pointer compression.
